src/Grpc.AspNetCore.HealthChecks/GrpcHealthChecksPublisher.cs

```diff
@@ -16,32 +16,86 @@
 
 #endregion
 
+using Grpc.Health.V1;
 using Grpc.HealthCheck;
 using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.Options;
 
 namespace Grpc.AspNetCore.HealthChecks;
 
 internal sealed class GrpcHealthChecksPublisher : IHealthCheckPublisher
 {
     private readonly HealthServiceImpl _healthService;
+    private readonly ILogger _logger;
     private readonly GrpcHealthChecksOptions _options;
 
-    public GrpcHealthChecksPublisher(HealthServiceImpl healthService, IOptions<GrpcHealthChecksOptions> options)
+    public GrpcHealthChecksPublisher(HealthServiceImpl healthService, IOptions<GrpcHealthChecksOptions> options, ILoggerFactory loggerFactory)
     {
         _healthService = healthService;
         _options = options.Value;
+        _logger = loggerFactory.CreateLogger<GrpcHealthChecksPublisher>();
     }
 
     public Task PublishAsync(HealthReport report, CancellationToken cancellationToken)
     {
-        foreach (var registration in _options.Services)
+        Log.EvaluatingPublishedHealthReport(_logger, report.Entries.Count, _options.Services.Count);
+
+        List<KeyValuePair<string, HealthReportEntry>>? serviceEntries = null;
+        foreach (var serviceMapping in _options.Services)
         {
-            var resolvedStatus = HealthChecksStatusHelpers.GetStatus(report, registration.Predicate);
+            serviceEntries ??= new();
+            serviceEntries.AddRange(report.Entries);
+
+            if (serviceMapping.HealthCheckPredicate != null)
+            {
+                for (var i = serviceEntries.Count - 1; i >= 0; i--)
+                {
+                    var entry = serviceEntries[i];
+                    var registration = new HealthCheckFilterContext(entry.Key, entry.Value.Tags);
+
+                    if (!serviceMapping.HealthCheckPredicate(registration))
+                    {
+                        serviceEntries.RemoveAt(i);
```

Comment:
`RemoveAt` copies the elements after `i` to be contiguous again. So if you'd like to avoid these copying maybe type the list as `List<(KeyValuePair<string, HealthReportEntry> Entry, bool Publish)>` and use the `Publish`-flag is indicator -- thus in the Linq a additional `.Where(e => e.Publish)` is needed too.